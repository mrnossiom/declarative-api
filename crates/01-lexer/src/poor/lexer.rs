use super::{
	cursor::Cursor,
	token::{
		DocStyle,
		LiteralKind::*,
		Token,
		TokenKind::{self, *},
	},
};
use unicode_xid::UnicodeXID;

impl<'a> Cursor<'a> {
	pub(crate) fn advance_token(&mut self) -> Token {
		let Some(first_char) = self.bump() else {
			return Token::new(Eof, 0);
		};

		let token_kind = match first_char {
			'#' => match self.first() {
				'#' | ' ' => self.line_comment(),
				_ => Pound,
			},

			// Whitespace sequence.
			c if is_whitespace(c) => self.whitespace(),

			// Identifier (this should be checked after other variant that can
			// start as identifier).
			c if is_id_start(c) => self.ident(),

			// Numeric literal.
			c @ '0'..='9' => self.number(c),

			// One-symbol tokens.
			';' => Semi,
			',' => Comma,
			'.' => Dot,
			'(' => OpenParenthesis,
			'{' => OpenBrace,
			'[' => OpenBracket,
			')' => CloseParenthesis,
			'}' => CloseBrace,
			']' => CloseBracket,
			'@' => At,
			'~' => Tilde,
			'?' => Question,
			':' => Colon,
			'$' => Dollar,
			'=' => Eq,
			'!' => Bang,
			'<' => Lt,
			'>' => Gt,
			'-' => Minus,
			'&' => And,
			'|' => Or,
			'+' => Plus,
			'*' => Star,
			'^' => Caret,
			'%' => Percent,

			// String literal.
			'"' => {
				let terminated = self.double_quoted_string();
				Literal(Str { terminated })
			}

			_ => Unknown,
		};

		let res = Token::new(token_kind, self.pos_within_token());
		self.reset_pos_within_token();
		res
	}

	fn line_comment(&mut self) -> TokenKind {
		debug_assert!(self.prev() == '#' && (self.first() == '#' || self.first() == ' '));
		self.bump();

		let style = match self.first() {
			'!' => Some(DocStyle::Inner),
			' ' => Some(DocStyle::Outer),
			_ => None,
		};

		self.eat_while(|c| c != '\n');
		LineComment(style)
	}

	fn whitespace(&mut self) -> TokenKind {
		debug_assert!(is_whitespace(self.prev()));
		self.eat_while(is_whitespace);
		Whitespace
	}

	fn ident(&mut self) -> TokenKind {
		debug_assert!(is_id_start(self.prev()));
		// Start is already eaten, eat the rest of identifier.
		self.eat_while(is_id_continue);
		// Known prefixes must have been handled earlier. So if
		// we see a prefix here, it is definitely an unknown prefix.
		match self.first() {
			c if !c.is_ascii() && unic_emoji_char::is_emoji(c) => self.fake_ident(),
			_ => Ident,
		}
	}

	fn fake_ident(&mut self) -> TokenKind {
		// Start is already eaten, eat the rest of identifier.
		self.eat_while(|c| {
			UnicodeXID::is_xid_continue(c)
				|| (!c.is_ascii() && unic_emoji_char::is_emoji(c))
				|| c == '\u{200d}'
		});

		InvalidIdent
	}

	/// Eats double-quoted string and returns true
	/// if string is terminated.
	fn double_quoted_string(&mut self) -> bool {
		debug_assert!(self.prev() == '"');
		while let Some(c) = self.bump() {
			match c {
				'"' => {
					return true;
				}
				'\\' if self.first() == '\\' || self.first() == '"' => {
					// Bump again to skip escaped character.
					self.bump();
				}
				_ => (),
			}
		}
		// End of file reached.
		false
	}

	fn number(&mut self, _c: char) -> TokenKind {
		self.eat_decimal_digits();

		Literal(Number)
	}
}

/// True if `c` is considered a whitespace
const fn is_whitespace(c: char) -> bool {
	// This is Pattern_White_Space.
	//
	// Note that this set is stable (ie, it doesn't change with different
	// Unicode versions), so it's ok to just hard-code the values.

	matches!(
		c,
		// Usual ASCII suspects
		'\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
	)
}

/// True if `c` is valid as a first character of an identifier.
fn is_id_start(c: char) -> bool {
	// This is XID_Start OR '_' (which formally is not a XID_Start).
	c == '_' || UnicodeXID::is_xid_start(c)
}

/// True if `c` is valid as a non-first character of an identifier.
fn is_id_continue(c: char) -> bool {
	UnicodeXID::is_xid_continue(c)
}

/// The passed string is lexically an identifier.
fn _is_ident(string: &str) -> bool {
	let mut chars = string.chars();
	chars.next().map_or(false, |start| {
		is_id_start(start) && chars.all(is_id_continue)
	})
}
