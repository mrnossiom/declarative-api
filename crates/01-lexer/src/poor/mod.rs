mod cursor;
mod lexer;
mod token;

pub use cursor::Cursor;
pub use token::*;
