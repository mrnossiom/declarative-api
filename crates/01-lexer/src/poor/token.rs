#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Token {
	pub kind: TokenKind,
	pub length: u32,
}

impl Token {
	pub(crate) const fn new(kind: TokenKind, length: u32) -> Self {
		Self { kind, length }
	}
}

impl Display for Token {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}", self.kind,)?;

		if !self.kind.is_single_char() {
			write!(f, " ({})", self.length)?;
		};

		Ok(())
	}
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
	// IDEA: maybe multiline
	LineComment(Option<DocStyle>),

	/// Any whitespace character.
	Whitespace,

	/// An identifier
	Ident,

	/// Like the above, but containing invalid unicode codepoints.
	InvalidIdent,

	/// See [LiteralKind] for more details.
	Literal(LiteralKind),

	// One-char tokens:
	/// ";"
	Semi,
	/// ","
	Comma,
	/// "."
	Dot,

	/// `(`
	OpenParenthesis,
	/// `)`
	CloseParenthesis,
	/// `{`
	OpenBrace,
	/// `}`
	CloseBrace,
	/// `[`
	OpenBracket,
	/// `]`
	CloseBracket,

	/// "@"
	At,
	/// "#"
	Pound,
	/// "~"
	Tilde,
	/// "?"
	Question,
	/// ":"
	Colon,
	/// "$"
	Dollar,
	/// "="
	Eq,
	/// "!"
	Bang,

	/// "<"
	Lt,
	/// ">"
	Gt,
	/// "-"
	Minus,
	/// "&"
	And,
	/// "|"
	Or,
	/// "+"
	Plus,
	/// "*"
	Star,
	/// "/"
	Slash,
	/// "^"
	Caret,
	/// "%"
	Percent,

	/// Unknown token, not expected by the lexer, e.g. "â„–"
	Unknown,

	/// End of input.
	Eof,
}

impl TokenKind {
	#[must_use]
	pub const fn is_whitespace(&self) -> bool {
		matches!(self, Self::Whitespace)
	}

	pub(crate) const fn is_eof(&self) -> bool {
		matches!(self, Self::Eof)
	}

	pub(crate) const fn is_single_char(&self) -> bool {
		!matches!(
			self,
			Self::LineComment(_)
				| Self::Whitespace
				| Self::Ident | Self::InvalidIdent
				| Self::Literal(_)
				| Self::Unknown
		)
	}
}

impl Display for TokenKind {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::LineComment(None) => write!(f, "line comment"),
			Self::LineComment(Some(DocStyle::Inner)) => write!(f, "inner line comment"),
			Self::LineComment(Some(DocStyle::Outer)) => write!(f, "outer line comment"),
			Self::Whitespace => write!(f, "whitespace"),

			Self::Ident => write!(f, "ident"),
			Self::InvalidIdent => write!(f, "invalid ident"),
			Self::Literal(LiteralKind::Number) => write!(f, "number literal"),
			Self::Literal(LiteralKind::Str { terminated: true }) => write!(f, "string literal"),
			Self::Literal(LiteralKind::Str { terminated: false }) => {
				write!(f, "non-terminated string literal")
			}

			Self::Semi => write!(f, ";"),
			Self::Comma => write!(f, ","),
			Self::Dot => write!(f, "."),

			Self::OpenParenthesis => write!(f, "("),
			Self::CloseParenthesis => write!(f, ")"),
			Self::OpenBrace => write!(f, "{{"),
			Self::CloseBrace => write!(f, "}}"),
			Self::OpenBracket => write!(f, "["),
			Self::CloseBracket => write!(f, "]"),

			Self::At => write!(f, "@"),
			Self::Pound => write!(f, "#"),
			Self::Tilde => write!(f, "~"),
			Self::Question => write!(f, "?"),
			Self::Colon => write!(f, ":"),
			Self::Dollar => write!(f, "$"),
			Self::Eq => write!(f, "="),
			Self::Bang => write!(f, "!"),

			Self::Lt => write!(f, ">"),
			Self::Gt => write!(f, "<"),
			Self::Minus => write!(f, "-"),
			Self::And => write!(f, "&"),
			Self::Or => write!(f, "|"),
			Self::Plus => write!(f, "+"),
			Self::Star => write!(f, "*"),
			Self::Slash => write!(f, "/"),
			Self::Caret => write!(f, "^"),
			Self::Percent => write!(f, "%"),

			Self::Unknown => write!(f, "unknown"),

			Self::Eof => write!(f, "<EOF>"),
		}
	}
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LiteralKind {
	/// "12_u8", "0o100", "0b120i99", "1f32".
	Number,
	/// ""abc"", ""abc"
	Str { terminated: bool },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AttrStyle {
	Outer,
	Inner,
	Inline,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DocStyle {
	// `##!` they document inside of scopes
	Inner,
	// `##` they document the item they embrace
	Outer,
}
