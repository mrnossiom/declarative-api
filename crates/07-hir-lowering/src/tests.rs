
use std::{
	str::FromStr,
	vec,
};


use super::*;

#[test]
fn test_1() {
	let v1 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};

	let v2 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};
	assert_eq!(v1, v2);
}

#[test]
fn test_2() {
	let v1 = types::Version {
		major: 1,
		minor: 2,
		patch: 3,
	};
	let v2 = types::Version {
		major: 0,
		minor: 3,
		patch: 4,
	};
	assert!(v1 > v2);
}



#[test]
#[allow(clippy::too_many_lines)]
fn test_3() {
	let input_ast = ast::types::Api {
		metadata: ast::types::ApiMetadata {
			name: Some("Wiro's API".into()),
			version: Some(ast::types::Version {
				major: 1,
				minor: 0,
				patch: 0,
			}),
			urls: vec![
				"https://paradigm.lighton.ai/api/v1".into(),
				"https://paradigm-preprod.lighton.ai/api/v1".into(),
				"https://paradigm-dev.lighton.ai/api/v1".into(),
			],
			comment: Some("Je suis un commentaire de documentation (des métadonnées)".into()),
		},
		data: ast::types::ApiData {
			child_scopes: vec![ast::types::Scope {
				name: "dashboard".into(),
				child_scopes: vec![],
				child_models: vec![ast::types::Model{
					name:"metrics".into(),
					model_body: vec![ast::types::KeyValuePair{
						key:"name".into(),
						type_:ast::types::Type::String,
						description:"The name of the metric".into(),
						parameters:vec![],
						comment:None,
					},ast::types::KeyValuePair{
						key:"email".into(),
						type_:ast::types::Type::String,
						description:String::new(),
						parameters:vec![],
						comment:None,
					},ast::types::KeyValuePair{
						key:"password".into(),
						type_:ast::types::Type::String,
						description:String::new(),
						parameters:vec![],
						comment:None,
					},
					],
				comment:None,
				}],
				child_paths: vec![ast::types::Path {
					name: "dashboard".into(),
					child_scopes: vec![],
					child_paths: vec![ast::types::Path {
						name: "metrics".into(),
						child_paths: vec![],
						comment: None,
						methods: vec![],
						headers: vec![ast::types::KeyValuePair {
							key: "Authorization".into(),
							type_: ast::types::Type::String,
							description: "The authorization key.".into(),
							parameters: vec![],
							comment: None,
						}],
						metadata: vec![],
						parameters: vec![],
						query: vec![],
						child_scopes: vec![ast::types::Scope {
							name: "metrics".into(),
							child_scopes: vec![],
							child_models: vec![],
							child_paths: vec![],
							methods: vec![ast::types::Method {
									method: http::Method::from_str("GET").expect(
												"a method used for testing the ast-lowering crate doesn't exist",
											),
											responses: vec![
												ast::types::Response{
													headers: vec![],
													status_code: 200,
													body:vec![
                                                        
                                                    ast::types::KeyValuePair{
														key:"metrics".into(),
														type_:ast::types::Type::Model("metrics".into()),
														description:String::new(),
														parameters:vec![],
														comment: None,
													},ast::types::KeyValuePair{
														key:"status".into(),
														type_:ast::types::Type::Object(
															vec![
																ast::types::KeyValuePair{
																	key:"message".into(),
																	type_:ast::types::Type::String,
																	description:"The status message itself".into(),
																	parameters:vec![],
																	comment: None,
																},ast::types::KeyValuePair{
																	key:"code".into(),
																	type_:ast::types::Type::Int,
																	description:"The status code".into(),
																	parameters:vec![],
																	comment: None,
																}
															]
														),
														description:"A status message container".into(),
														parameters:vec![],
														comment: None,
													}],
													comment: None
												}
											],
											headers: vec![],
											query: vec![],
											comment: None,
										}],
							parameters: vec![],
							headers: vec![],
							query: vec![],
							comment: None,
						}],
					}],
					methods: vec![],
					headers: vec![],
					parameters: vec![],
					query: vec![],
					metadata: vec![],
					comment: None,
				}],
				methods: vec![],
				parameters: vec![],
				headers: vec![],
				query: vec![],
				comment: None,
			}],
		},
	};
	
	
	let mut parsed_intermediate_representation = types::IntermediateRepresentation::default();
	input_ast.parse(
		types::ParserVariables::default(),
		&mut parsed_intermediate_representation,
	);
	let expected_intermediate_representation = utils::gen_test_ir();
	//dbg!(&expected_intermediate_representation);
	//dbg!(&parsed_intermediate_representation);
	assert!(expected_intermediate_representation == parsed_intermediate_representation);
}
