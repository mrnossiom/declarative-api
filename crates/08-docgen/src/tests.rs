use super::*;
use hir_lowering::{utils, Parse};

#[test]
fn test_1() {
	let test_ir = utils::gen_test_ir();
	let mut generated_file_list = HashMap::new();
	test_ir.markdown(&mut generated_file_list);
	// generated_file_list.write(Path::new("test-artifacts/test-1"));
}

#[test]
#[allow(clippy::too_many_lines)]
fn test_2() {
	let input_hir = hir::types::Api {
		metadata: hir::types::ApiMetadata {
			name: Some("Wiro's API".into()),
			version: Some(hir::types::Version {
				major: 1,
				minor: 0,
				patch: 0,
			}),
			urls: vec![
				"https://paradigm.lighton.ai/api/v1".into(),
				"https://paradigm-preprod.lighton.ai/api/v1".into(),
				"https://paradigm-dev.lighton.ai/api/v1".into(),
			],
			comment: Some("This is the API of Wiro".into()),
		},
		data: hir::types::ApiData {
			scopes: vec![hir::types::Scope {
				name: "dashboard".into(),
				models: vec![],
				scopes: vec![],
				methods: vec![],
				paths: vec![hir::types::Path {
					name: "dashboard".into(),
					scopes: vec![],
					paths: vec![],
					methods: vec![],
					headers: vec![
						hir::types::KeyValuePair {
							key: "Authorization".into(),
							type_: hir::types::Type::String,
							description: "The API Key of the User".into(),
							parameters: vec![("prefix".into(), "Api-Key".into())],
							comment: None,
						},
						hir::types::KeyValuePair {
							key: "X-Model".into(),
							type_: hir::types::Type::String,
							description: "The Model of the User".into(),
							parameters: vec![],
							comment: None,
						},
					],
					parameters: vec![],
					query: vec![],
					metadata: vec![],
					comment: None,
				}],
				headers: vec![],
				parameters: vec![],
				query: vec![],
				comment: None,
			}],
		},
	};
	let mut parsed_intermediate_representation = hir_lowering::types::Ir::default();
	input_hir.parse(
		hir_lowering::types::ParserVariables::default(),
		&mut parsed_intermediate_representation,
	);
	let mut generated_file_list = HashMap::new();
	parsed_intermediate_representation.markdown(&mut generated_file_list);
	// generated_file_list.write(Path::new("test-artifacts/test-2"));
}
