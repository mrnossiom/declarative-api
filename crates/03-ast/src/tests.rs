use crate::types::*;
use lexer::{span::Span, symbols::Symbol};

#[allow(clippy::too_many_lines)]
fn _dummy_api() -> Api {
	Api {
		meta: vec![
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This Api is a test for the documentation generator of the api".into(),
				span: Span { start: 0, end: 0 },
			},
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This is a second line of comment".into(),
				span: Span { start: 0, end: 0 },
			},
			Attribute {
				ident: Ident {
					symbol: Symbol::new_static("doc"),
					span: Span { start: 0, end: 0 },
				},
				value: " This is a third line of comment".into(),
				span: Span { start: 0, end: 0 },
			},
		],
		items: vec![
			Item {
				attrs: vec![],
				kind: ItemKind::Meta(Meta {
					fields: vec![
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("name"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("Wiro's API".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("description"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("This is the API of Wiro".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("version"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Str("1.0.0".into()),
							span: Span { start: 0, end: 0 },
						},
						MetaField {
							ident: Ident {
								symbol: Symbol::new_static("name"),
								span: Span { start: 0, end: 0 },
							},
							value: MetaFieldKind::Vec(vec![
								MetaFieldKind::Str("https://paradigm.lighton.ai/api/v1".into()),
								MetaFieldKind::Str(
									"https://paradigm-preprod.lighton.ai/api/v1".into(),
								),
								MetaFieldKind::Str("https://paradigm-dev.lighton.ai/api/v1".into()),
							]),
							span: Span { start: 0, end: 0 },
						},
					],
				}),
				span: Span { start: 0, end: 0 },
				ident: Ident {
					symbol: Symbol::new_static(""),
					span: Span { start: 0, end: 0 },
				},
			},
			Item {
				attrs: vec![Attribute {
					ident: Ident {
						symbol: Symbol::new_static("doc"),
						span: Span { start: 0, end: 0 },
					},
					value: " Imports the `builder.dapi` file".into(),
					span: Span { start: 0, end: 0 },
				}],
				ident: Ident {
					symbol: Symbol::new_static("builder"),
					span: Span { start: 0, end: 0 },
				},
				kind: ItemKind::Scope(ScopeKind::Unloaded),
				span: Span { start: 0, end: 0 },
			},
			Item {
				attrs: vec![],
				ident: Ident {
					symbol: Symbol::new_static("dashboard"),
					span: Span { start: 0, end: 0 },
				},
				kind: ItemKind::Scope(ScopeKind::Loaded {
					items: vec![Item {
						attrs: vec![],
						ident: Ident {
							symbol: Symbol::new_static("dashboard"),
							span: Span { start: 0, end: 0 },
						},
						kind: ItemKind::Path(Path {
							path: PathKind::String(Ident {
								symbol: Symbol::new_static("dashboard"),
								span: Span { start: 0, end: 0 },
							}),
							items: vec![Item {
								attrs: vec![],
								ident: Ident {
									symbol: Symbol::new_static(""),
									span: Span { start: 0, end: 0 },
								},
								kind: ItemKind::Headers(Headers {
									headers: vec![
										HeaderField {
											attrs: vec![
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " # Safety".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " This is a comment".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("doc"),
														span: Span { start: 0, end: 0 },
													},
													value: " This is a second line of comment"
														.into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("description"),
														span: Span { start: 0, end: 0 },
													},
													value: "The API Key of the User".into(),
													span: Span { start: 0, end: 0 },
												},
												Attribute {
													ident: Ident {
														symbol: Symbol::new_static("prefix"),
														span: Span { start: 0, end: 0 },
													},
													value: "Api-Key".into(),
													span: Span { start: 0, end: 0 },
												},
											],
											ident: Ident {
												symbol: Symbol::new_static("Authorization"),
												span: Span { start: 0, end: 0 },
											},
											span: Span { start: 0, end: 0 },
										},
										HeaderField {
											attrs: vec![Attribute {
												ident: Ident {
													symbol: Symbol::new_static("description"),
													span: Span { start: 0, end: 0 },
												},
												value: "The Model of the User".into(),
												span: Span { start: 0, end: 0 },
											}],
											ident: Ident {
												symbol: Symbol::new_static("X-Model"),
												span: Span { start: 0, end: 0 },
											},
											span: Span { start: 0, end: 0 },
										},
									],
								}),
								span: Span { start: 0, end: 0 },
							}],
						}),
						span: Span { start: 0, end: 0 },
					}],
					span: Span { start: 0, end: 0 },
				}),
				span: Span { start: 0, end: 0 },
			},
		],
		span: Span { start: 0, end: 0 },
	}
}
