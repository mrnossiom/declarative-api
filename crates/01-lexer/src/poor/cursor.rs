use crate::poor::Token;
use std::str::Chars;

pub struct Cursor<'a> {
	len_remaining: usize,
	chars: Chars<'a>,

	#[cfg(debug_assertions)]
	prev: char,
}

pub(crate) const EOF_CHAR: char = '\0';

impl<'a> Cursor<'a> {
	#[must_use]
	pub fn new(source: &'a str) -> Cursor<'a> {
		Cursor {
			len_remaining: source.len(),
			chars: source.chars(),
			#[cfg(debug_assertions)]
			prev: EOF_CHAR,
		}
	}

	/// Returns the last eaten symbol.
	#[cfg(debug_assertions)]
	pub(super) const fn prev(&self) -> char {
		self.prev
	}

	/// Peeks the next symbol from the input stream without consuming it.
	/// If requested position doesn't exist, `EOF_CHAR` is returned.
	/// However, getting `EOF_CHAR` doesn't always mean actual end of file,
	/// it should be checked with `is_eof` method.
	pub(super) fn first(&self) -> char {
		self.chars.clone().next().unwrap_or(EOF_CHAR)
	}

	/// Peeks the second symbol from the input stream without consuming it.
	pub(super) fn _second(&self) -> char {
		let mut iter = self.chars.clone();
		iter.next();
		iter.next().unwrap_or(EOF_CHAR)
	}

	/// Checks if there is nothing more to consume.
	pub(super) fn is_eof(&self) -> bool {
		self.chars.as_str().is_empty()
	}

	/// Returns amount of already consumed symbols.
	pub(super) fn pos_within_token(&self) -> u32 {
		(self.len_remaining - self.chars.as_str().len()) as u32
	}

	/// Resets the number of bytes consumed to 0.
	pub(super) fn reset_pos_within_token(&mut self) {
		self.len_remaining = self.chars.as_str().len();
	}

	/// Moves to the next character.
	pub(super) fn bump(&mut self) -> Option<char> {
		let c = self.chars.next()?;

		#[cfg(debug_assertions)]
		{
			self.prev = c;
		}

		Some(c)
	}

	/// Eats symbols while predicate returns true or until the end of file is reached.
	pub(super) fn eat_while(&mut self, mut predicate: impl FnMut(char) -> bool) {
		while predicate(self.first()) && !self.is_eof() {
			self.bump();
		}
	}
}

pub struct Iter<'a>(Cursor<'a>);

impl<'a> Iterator for Iter<'a> {
	type Item = Token;

	fn next(&mut self) -> Option<Self::Item> {
		let token = self.0.advance_token();

		if token.kind.is_eof() {
			None
		} else {
			Some(token)
		}
	}
}

impl<'a> IntoIterator for Cursor<'a> {
	type Item = Token;
	type IntoIter = Iter<'a>;

	fn into_iter(self) -> Self::IntoIter {
		Iter(self)
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::tests::{ATTR, EXAMPLE};

	#[test]
	fn can_tokenize_example_file() {
		let _poor_tokens = Cursor::new(EXAMPLE).into_iter().collect::<Vec<_>>();
	}

	#[test]
	fn parse_attr_style() {
		use crate::poor::TokenKind::*;

		let mut tokens = Cursor::new(ATTR).into_iter();

		macro_rules! next_token {
			($ty:ident, $len:literal) => {
				assert_eq!(tokens.next(), Some(Token::new($ty, $len)));
			};
			(@end) => {
				assert_eq!(tokens.next(), None);
			};
		}

		next_token!(At, 1);
		next_token!(Ident, 6);
		next_token!(Colon, 1);
		next_token!(Whitespace, 1);
		next_token!(Ident, 4);
		next_token!(@end);
	}
}
