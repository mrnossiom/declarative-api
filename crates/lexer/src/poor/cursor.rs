use crate::poor::Token;
use std::str::Chars;

pub struct Cursor<'a> {
	len_remaining: usize,
	chars: Chars<'a>,

	#[cfg(debug_assertions)]
	prev: char,
}

pub(crate) const EOF_CHAR: char = '\0';

impl<'a> Cursor<'a> {
	#[must_use]
	pub fn from_source(source: &'a str) -> Cursor<'a> {
		Cursor {
			len_remaining: source.len(),
			chars: source.chars(),
			#[cfg(debug_assertions)]
			prev: EOF_CHAR,
		}
	}

	/// Returns the last eaten symbol.
	#[cfg(debug_assertions)]
	pub(super) const fn prev(&self) -> char {
		self.prev
	}

	/// Peeks the next symbol from the input stream without consuming it.
	/// If requested position doesn't exist, `EOF_CHAR` is returned.
	/// However, getting `EOF_CHAR` doesn't always mean actual end of file,
	/// it should be checked with `is_eof` method.
	pub(super) fn first(&self) -> char {
		self.chars.clone().next().unwrap_or(EOF_CHAR)
	}

	/// Peeks the second symbol from the input stream without consuming it.
	pub(super) fn _second(&self) -> char {
		let mut iter = self.chars.clone();
		iter.next();
		iter.next().unwrap_or(EOF_CHAR)
	}

	/// Checks if there is nothing more to consume.
	pub(super) fn is_eof(&self) -> bool {
		self.chars.as_str().is_empty()
	}

	/// Returns amount of already consumed symbols.
	///
	/// # Panics
	/// When used in a file that is over 4GiB
	pub(super) fn pos_within_token(&self) -> u32 {
		u32::try_from(self.len_remaining - self.chars.as_str().len())
			.expect("loaded sources can't go over 4 GiB")
	}

	/// Resets the number of bytes consumed to 0.
	pub(super) fn reset_pos_within_token(&mut self) {
		self.len_remaining = self.chars.as_str().len();
	}

	/// Moves to the next character.
	pub(super) fn bump(&mut self) -> Option<char> {
		let c = self.chars.next()?;

		#[cfg(debug_assertions)]
		{
			self.prev = c;
		}

		Some(c)
	}

	/// Eats symbols while predicate returns true or until the end of file is reached.
	pub(super) fn eat_while(&mut self, mut predicate: impl FnMut(char) -> bool) {
		while predicate(self.first()) && !self.is_eof() {
			self.bump();
		}
	}

	pub(super) fn eat_decimal_digits(&mut self) -> bool {
		let mut has_digits = false;
		loop {
			match self.first() {
				'_' => {
					self.bump();
				}
				'0'..='9' => {
					has_digits = true;
					self.bump();
				}
				_ => break,
			}
		}
		has_digits
	}
}

pub struct Iter<'a>(Cursor<'a>);

impl<'a> Iterator for Iter<'a> {
	type Item = Token;

	fn next(&mut self) -> Option<Self::Item> {
		let token = self.0.next_token();

		if token.kind.is_eof() {
			None
		} else {
			Some(token)
		}
	}
}

impl<'a> IntoIterator for Cursor<'a> {
	type Item = Token;
	type IntoIter = Iter<'a>;

	fn into_iter(self) -> Self::IntoIter {
		Iter(self)
	}
}

#[cfg(test)]
mod tests {
	use crate::tests::{ATTR, EXAMPLE, URLS};

	macro_rules! assert_tokenize {
		($variant:literal, $src:ident) => {
			paste::paste! {
				#[test]
				fn [<tokenize_ $variant>]() {
					let vec = super::Cursor::from_source($src)
						.into_iter()
						.collect::<Vec<_>>();
					insta::assert_debug_snapshot!(vec);
				}
			}
		};
	}

	assert_tokenize!("example", EXAMPLE);
	assert_tokenize!("attribute", ATTR);
	assert_tokenize!("urls", URLS);
}
