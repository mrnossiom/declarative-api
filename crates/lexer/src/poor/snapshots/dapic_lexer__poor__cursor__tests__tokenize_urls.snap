---
source: crates/lexer/src/poor/cursor.rs
expression: vec
---
[
    Token {
        kind: Ident,
        length: 4,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: OpenBracket,
        length: 1,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 36,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 44,
    },
    Token {
        kind: Whitespace,
        length: 2,
    },
    Token {
        kind: Literal(
            Str {
                terminated: true,
            },
        ),
        length: 40,
    },
    Token {
        kind: Whitespace,
        length: 1,
    },
    Token {
        kind: CloseBracket,
        length: 1,
    },
]
